### PR TITLE
fix: add list of the props, which need to be placed at the root

### PR DESCRIPTION
**Pull request by ItsSudip**
## What are the changes introduced in this PR?

Adding a new config field called `listOfProps` to select a list of properties needs to be placed on the root of the eventValue.

## What is the related Linear task?

Resolves INT-2733

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3777/test-report.html

File: test/integrations/destinations/af/processor/data.ts [hunk 1]
```diff
@@ -1716,6 +1716,98 @@ export const existingTestCases = [
...
+              event: 'Order Completed',
+              sentAt: '2020-08-14T05:30:30.118Z',
+              context: commonContextWithExternalId,
+              messageId: '7208bbb6-2c4e-45bb-bf5b-ad426f3593e9',
+              timestamp: '2020-08-14T05:30:30.118Z',
+              properties: {
+                ...commonPropertiesWithProduct,
+                prop1: 'value1',
+                prop2: 'value2',
+                Prop3: 'value3',
```
Comment (krishna2020) [inline test/integrations/destinations/af/processor/data.ts:1741]:
```suggestion
                prop3: 'value3',
```

File: test/integrations/destinations/af/processor/data.ts [hunk 2]
```diff
@@ -1716,6 +1716,98 @@ export const existingTestCases = [
...
+                trackRevenuePerProduct: false,
+                addPropertiesAtRoot: false,
+                listOfProps: [
+                  {
+                    property: 'prop1',
+                  },
+                  {
+                    property: 'prop2',
+                  },
+                  {
```
Comment (krishna2020) [inline test/integrations/destinations/af/processor/data.ts:1762]: remove prop3 from here to cover 2 scenarios of being at root vs. properties

File: test/integrations/destinations/af/processor/data.ts (hunk 1 above)
Comment (ItsSudip) [reply test/integrations/destinations/af/processor/data.ts:1741]: This is intensional to check case sensitivity.

File: test/integrations/destinations/af/processor/data.ts (hunk 2 above)
Comment (ItsSudip) [reply test/integrations/destinations/af/processor/data.ts:1762]: This is also intensional to check the case sensitivity.

Comment (krishna2020) [reply test/integrations/destinations/af/processor/data.ts:1762]: In that case, Prop3 should not be at root level ?

**Review by krishna2020 — DISMISSED** [review]

**Review by Gauravudia — APPROVED** [review]

**Review by shrouti1507 — DISMISSED** [review]

**Review by shrouti1507 — DISMISSED** [review]

**Review by yashasvibajpai — DISMISSED** [review]

File: src/v0/destinations/af/transform.js [hunk 3]
```diff
@@ -130,28 +130,29 @@ function getEventValueForUnIdentifiedTrackEvent(message) {
 
 function getEventValueMapFromMappingJson(message, mappingJson, isMultiSupport, config) {
   let eventValue = {};
-  const { addPropertiesAtRoot, afCurrencyAtRoot } = config;
-
+  const { addPropertiesAtRoot, afCurrencyAtRoot, listOfProps } = config;
+  const clonedProp = message.properties && JSON.parse(JSON.stringify(message.properties));
```
Comment (koladilip) [inline src/v0/destinations/af/transform.js:134]: why not use lodash cloneDeep?

File: src/v0/destinations/af/transform.js [hunk 4]
```diff
@@ -130,28 +130,29 @@ function getEventValueForUnIdentifiedTrackEvent(message) {
...
   if (addPropertiesAtRoot) {
-    eventValue = message.properties;
+    eventValue = clonedProp;
   } else {
-    set(eventValue, 'properties', message.properties);
+    if (Array.isArray(listOfProps) && listOfProps.length > 0) {
+      listOfProps.forEach((prop) => {
```
Comment (koladilip) [inline src/v0/destinations/af/transform.js:139]: if prop.property is simple string not containing any dots then why use set function?
instead we can set using eventValue[prop.property] = message.properties[properties] or we can use lodash pick method and pass array

File: src/v0/destinations/af/transform.js (hunk 3 above)
Comment (ItsSudip) [reply src/v0/destinations/af/transform.js:134]: addressed

File: src/v0/destinations/af/transform.js (hunk 4 above)
Comment (ItsSudip) [reply src/v0/destinations/af/transform.js:139]: addressed

File: src/v0/destinations/af/transform.js [hunk 5]
```diff
@@ -130,28 +130,29 @@ function getEventValueForUnIdentifiedTrackEvent(message) {
...
+        eventValue[prop.property] = clonedProp[prop.property];
+        delete clonedProp[prop.property];
+      });
+    }
+    set(eventValue, 'properties', clonedProp);
```
Comment (koladilip) [inline src/v0/destinations/af/transform.js:144]: this can be avoid

File: src/v0/destinations/af/transform.js [hunk 6]
```diff
@@ -130,28 +130,29 @@ function getEventValueForUnIdentifiedTrackEvent(message) {
...
   }
 
   const sourceKeys = Object.keys(mappingJson);
   sourceKeys.forEach((sourceKey) => {
     set(eventValue, mappingJson[sourceKey], get(message, sourceKey));
```
Comment (koladilip) [inline src/v0/destinations/af/transform.js:149]: eventValue[mappingJson[sourceKey]] =  message[sourceKey];

mappingJson[sourceKey] does it contain complex values like a.b.c?

Comment (ItsSudip) [reply src/v0/destinations/af/transform.js:149]: I am not sure of this.

File: src/v0/destinations/af/transform.js (hunk 5 above)
Comment (ItsSudip) [reply src/v0/destinations/af/transform.js]: addressed

**Review by koladilip — APPROVED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Pull request merged**